[FLOC-2647] Documentation explaining how to export Flocker log files

wallrj opened the pull request:
Fixes: https://clusterhq.atlassian.net/browse/FLOC-2647

jongiddy commented on docs/using/administering/debugging.rst:
```diff
@@ -14,6 +14,35 @@ Ubuntu
+The relevant logs can be exported by running the following commands as root:
+
+# FLOC-2647 Perhaps this would be better as a standalone script.
+# Followup issue or in this branch?
```
Given the complexity and multiple possibilities for failures, I think it would be better as a script.

jongiddy commented on docs/using/administering/debugging.rst:
```diff
@@ -14,6 +14,35 @@ Ubuntu
+   # Export the full syslog since last boot with UTC timestamps
+   gzip < /var/log/syslog > "${ARCHIVE_NAME}/all-${SUFFIX}.log.gz"
+   # Create a single archive file
+   tar --create --file "${ARCHIVE_NAME}.tar" "${ARCHIVE_NAME}"
```
You should get better compression by tarring first and then gzipping the whole tar file (lots of shared patterns between the different log files).

However, if possible, it would also be good to avoid creating the whole tar file on disk.

This may only be tractable by using a script, or using a `zip` file.

jongiddy commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +66,37 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+# FLOC-2647 Perhaps this would be better as a standalone script.
+# Followup issue or in this branch?
+.. prompt:: bash [root@node1]# auto
```
Ditto on use of script and compressing after archiving.

jongiddy commented on docs/using/administering/debugging.rst:
```diff
@@ -23,6 +52,10 @@ They have unit names starting constructed with a ``flocker-`` prefix, e.g. ``flo
 
 It is possible to see the available unit names, and then view the logs with ``journalctl``:
 
+# FLOC-2647 Not sure if this example is helpful anymore
```
Leave it here as a first step for monitoring Flocker, before getting to the stage of needing to report a bug. Although there is some repetition, a user might not realise that a section entitled Bug Reporting contains information that is useful when not intending to report a bug.

You can however change the first command to gather the unit names.

jongiddy commented on CONTRIBUTING.rst:
```diff
@@ -41,6 +41,9 @@ Talk to Us
 Have questions or need help?
 
+# FLOC-2647 Maybe add a separate section which describes an ideal issue report.
+# And link to the log export documentation in debugging.rst?
```
I think a link here would be useful. I had trouble getting to the Debugging page (`using/administering/debugging.html`) from the page into which this file gets integrated (`gettinginvolved/contributing`).

:question: File an issue for @davidoliverSP2 to review the content and relative position of these two pages.

jongiddy commented on CONTRIBUTING.rst:
```diff
@@ -41,6 +41,9 @@ Talk to Us
 
 Have questions or need help?
 
+# FLOC-2647 Maybe add a separate section which describes an ideal issue report.
```
This would be better placed on the Debugging page.

wallrj replied:
Done.

jongiddy commented on docs/using/administering/debugging.rst:
```diff
@@ -14,6 +14,35 @@ Ubuntu
+Ubuntu Bug Reporting
+^^^^^^^^^^^^^^^^^^^^
+
+When reporting issues with Flocker on Ubuntu always include copies of all the log files.
```
:question:  "always" -> "please", here and in CentOS section.

jongiddy commented:
Thanks. Please proceed with coding. I think you should start with scripts, one for each OS, and perhaps make a follow-up issue for a `flocker-diagnostics` tool.

jml commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +35,46 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+When reporting issues with Flocker please include:
+
+* Operating System and version,
```
I don't know our house style, but probably "Operating System" should be lower case. Might be good to add `(e.g. "Ubuntu 15.04")`.

davidoliverSP2 replied:
Yep, we only use title case in our headers - in this context it should be Operating system

jml replied:
For my own education, does that mean it should also be "Kernel version" and "All Flocker log files" in the bullet points below?

jml commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +35,46 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+.. literalinclude:: flocker-log-export-ubuntu.sh
+   :language: sh
+
```
Is there a way I can see a preview of this?

wallrj replied:
You click on the `docs` build results below and then look through the build steps for `upload-html` and click on `build-9098`:

* http://build.clusterhq.com/results/docs/log-documentation-FLOC-2647-2/build-9098/using/administering/debugging.html#bug-reporting

jml commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +35,46 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+The following scripts can be used to gather this information on Ubuntu and CentOS.
+
+Save the appropriate script to a file and then run it as root
```
"as root." – Please add a full stop.

davidoliverSP2 replied:
I would prefer this instruction to appear between the download file, and the prompt in each of the sub-sections:

For example, 
Save the script to a file and then run the following as root:

jml commented on docs/using/administering/flocker-log-export-centos.sh:
```diff
@@ -0,0 +1,30 @@
+
+# Create a single archive file
+tar --create --file "${ARCHIVE_NAME}.tar" "${ARCHIVE_NAME}"
+rm -rf "${ARCHIVE_NAME}"
```
Although I don't like the way he said it, @exarkun is right that we would be better having untested Python code than untested shell code. That would also help reduce the duplication, allowing us to provide one "universal" script, which is probably a better user experience.

Some of the shell commands here could then be factored back into the documentation. e.g. `* kernel version (output of uname -a)`

jml commented:
Thanks! This looks like it will be a huge help to us and to our users.

Please make the requested changes and bounce back to me for review.

wallrj commented:
Ready for another review. I've:
 * replaced the shell scripts with a single downloadable python script and
 * moved the shell commands into the debugging documentation as an alternative way of gathering the logs and debug info in case the script doesn't work on the node OS.

itamarst commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +37,50 @@ It is possible to see the available unit names, and then view the logs with ``jo
+* operating system and version,
+* Linux kernel version,
+* Docker version and configuration,
+* Flocker log files and
```
Flocker version too.

itamarst commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +37,50 @@ It is possible to see the available unit names, and then view the logs with ``jo
+* Linux kernel version,
+* Docker version and configuration,
+* Flocker log files and
+* recent syslog content.
```
`syslog` is inaccurate; on Ubuntu we write our log files out. Maybe "Recent Flocker log content"?

wallrj replied:
Well we include Flocker log content separately. We want the entire syslog for kernel  and Docker log messages etc.

itamarst commented on docs/using/administering/flocker-log-export.py:
```diff
@@ -0,0 +1,319 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
```
ClusterHQ!

itamarst commented:
That's a pretty useful-looking tool, and I expect it will save us lots of time debugging user installs.

However, having to manually download it seems like an extra step that will make it harder for people to use it, and result in potential version mismatches etc.. Additionally, lacking tests we don't know that this will continue to work.

I therefore propose:
* Move the code into the `flocker` Python package.
* Add entry point binary that calls it to the `clusterhq-flocker-node` RPMs/DEBs.
* Write a minimal acceptance test that demonstrates it collects the expected information. No need for unit tests or functional tests, acceptance test should demonstrate it does what we need.

The docs are then "run the pre-installed `flocker-bug-report-dump`" or whatever it's called, so one less step.

itamarst commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +37,50 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+The following script can be used to gather this information on Ubuntu 14.04 and CentOS 7.
+
+* :download:`flocker-log-export.py`
```
It doesn't just export logs, so give it name that reflects that. `flocker-bug-report-dump` or something?

adamtheturtle replied:
diagnostics report?

wallrj replied:
`flocker-diagnostics`

itamarst commented:
Oh, and maybe docs should note that resulting log dump might contain sensitive information so they should be careful before sending it to us.

Address the above and resubmit - this is going to be super-useful, thanks for taking it on.

wallrj commented:
Ready for another review:
* Added ``flocker-diagnostics`` to ``clusterhq-flocker-node`` 
* Archive separate startup logs and eliot log
* Added an acceptance test that runs the tool and checks the contents of the resulting file.
* Changed the archive file name to have a uuid1 suffix rather than timestamp and hostname.

jml commented on flocker/node/_diagnostics.py:
```diff
@@ -0,0 +1,313 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
```
Who's Hybrid Logic?

wallrj replied:
Copied and pasted from the nearest neighbouring module. I'll change it.

jml commented on flocker/node/_diagnostics.py:
```diff
@@ -0,0 +1,313 @@
+def gzip_file(source_path, archive_path):
+    """
+    Create a gzip compressed archive of ``source_path`` at ``archive_path``.
+    An empty archive file will be created if the source file does not exist.
```
Why is this desirable behaviour?

wallrj replied:
It's bad I know. 

My reason is that there won't be any `flocker-*_startup` logs on Ubuntu unless there are (or have been) errors during service startup.
I wanted a consistent set of files in the diagnostic archive for testing purposes.

jml replied:
Could you please add a note along those lines, e.g. 

"… does not exist, so we have a consistent set of files in the diagnostic archive, which makes testing easier."

jml replied:
Done.

jml commented on flocker/acceptance/endtoend/test_diagnostics.py:
```diff
@@ -0,0 +1,100 @@
+from ..testtools import require_cluster
+
+# XXX: Copied from ``flocker.admin.acceptance``.
+# Should probably refactor and re-use ``flocker.admin.runner`` too.
```
:question: Changing `admin.acceptance` to use this list would be a very small change. I encourage you to do it in this patch, because otherwise it's unlikely to happen.

jml commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +37,78 @@ It is possible to see the available unit names, and then view the logs with ``jo
+* node IP addresses,
+* node hostname,
+* recent syslog content,
+* and separate Flocker log files.
```
:question: In our recent conversations when helping some users, we've also requested the following outputs:

* `cinder list`
* `nova volume-list`
* `lsblk -a`
* `fdisk -l`

I would definitely recommend **not** doing the openstack stuff as part of this PR.

I am less sure about whether we should include the other two. On the one hand, we know we need them, on the other hand, this PR is already a huge improvement on the current situation, so I'm inclined to get it "in the bank", so to speak.

If you do add `lsblk -a` and `fdisk -l`, please resubmit for further review.

davidoliverSP2 replied:
It might make this a little easier to read by using complete sentences in the bulleted list:

* The version of Flocker you are using.
* Your operating system and version.
* Your Linux kernel version.
* The version of Docker you are using, and your configuration details.
* Your node IP addresses.
* Your node hostname.
* All recent syslog content.
* Any separate Flocker log files.

wallrj replied:
Done.

wallrj replied:
> If you do add lsblk -a and fdisk -l, please resubmit for further review.

They seem pretty important so I added them.

jml commented on flocker/node/_diagnostics.py:
```diff
@@ -0,0 +1,313 @@
+        open(target_path + '_startup.gz', 'w').close()
+        check_call(
+            'journalctl --all --output cat --unit {}.service '
+            '| gzip'.format(service_name),
```
Why did you take the approach of gzipping individual files and then tarring them, rather than the other way around, which is more traditional?

wallrj replied:
I was trying to reduce the amount of disk space used by the tool. 
We know that our log files are huge and nodes have been known to run out of disk space, so I tried not to make things worse.

Ideally the `tarfile` or `zipfile` modules would give you a file like object that could be supplied to `check_call(stdout=...` but there doesn't seem to be such an interface.

Nor could I find an option for the `tar` command to append the content of stdin as a named file in the archive.

jml replied:
Ack

jml commented:
Thanks Richard, this is looking great. I think the patch is broadly fine to go in as-is. I've asked a couple of fairly open-ended questions. If the answers don't involve changed code, then you should feel free to merge once the tests pass, and once you've dealt with the minor changes that I've flagged. 

I think I've confused myself a little, so please ping me if I've confused you.

wallrj commented on admin/packaging.py:
```diff
@@ -910,6 +910,8 @@ def omnibus_package_builder(
                     (FilePath('/opt/flocker/bin/flocker-dataset-agent'),
                      flocker_node_path),
+                    (FilePath('/opt/flocker/bin/flocker-diagnostics'),
+                     flocker_node_path),
```
@jml These are the stupid tests we were discussing yesterday....I could remove them rather than making them uglier. What do you think?

jml replied:
I think I addressed this in a previous pass. Commenting now for completeness & ease of tracking.

wallrj commented on admin/test/test_packaging.py:
```diff
@@ -973,6 +973,8 @@ def fake_make_dependencies(
                         (FilePath('/opt/flocker/bin/flocker-dataset-agent'),
                          flocker_node_path),
+                        (FilePath('/opt/flocker/bin/flocker-diagnostics'),
+                         flocker_node_path),
```
...and here.

jml replied:
Commented in previous pass.

wallrj commented on flocker/acceptance/endtoend/test_diagnostics.py:
```diff
@@ -0,0 +1,100 @@
+            'service-status',
+            'ip-addr',
+            'hostname',
+        ]
```
And speaking of stupid tests....what's your honest opinion of this one? 
I thought about attempting to trigger some interesting flocker-control and flocker-dataset-agent log messages and then check that they are contained in the diagnostic bundle, but it seemed too difficult / not worth the effort. Is it worth doing now?

wallrj commented:
Aha, I see the problem:

 * http://build.clusterhq.com/builders/flocker%2Facceptance%2Faws%2Fubuntu-14.04%2Faws/builds/1676/steps/trial/logs/test.log
```
2015-08-13 14:03:01+0000 [-] ELIOT: {"task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "command": ["scp", "-C", "-q", "-o", "StrictHostKeyChecking=no", "-o", "ControlMaster=no", "-o", "PreferredAuthentications=publickey", "root@54.67.122.218:/srv/buildslave/flocker-acceptance-aws-ubuntu-14.04-aws/tmp-1676/_trial_temp/WARNING: No swap limit support", "/srv/buildslave/flocker-acceptance-aws-ubuntu-14.04-aws/tmp-1676/_trial_temp/flocker.acceptance.endtoend.test/DiagnosticsTests/test_export/dr725_/temp"], "task_level": [1], "action_type": "flocker.common.runner:run", "timestamp": 1439474581.996801, "action_status": "started"}
2015-08-13 14:03:02+0000 [-] ELIOT: {"timestamp": 1439474582.502729, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "message_type": "flocker.common.runner:run:output", "line": "scp: /srv/buildslave/flocker-acceptance-aws-ubuntu-14.04-aws/tmp-1676/_trial_temp/WARNING:: No such file or directory", "task_level": [2]}
2015-08-13 14:03:02+0000 [-] ELIOT: {"timestamp": 1439474582.503312, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "message_type": "flocker.common.runner:run:output", "line": "scp: No: No such file or directory", "task_level": [3]}
2015-08-13 14:03:02+0000 [-] ELIOT: {"timestamp": 1439474582.503575, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "message_type": "flocker.common.runner:run:output", "line": "scp: swap: No such file or directory", "task_level": [4]}
2015-08-13 14:03:02+0000 [-] ELIOT: {"timestamp": 1439474582.503801, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "message_type": "flocker.common.runner:run:output", "line": "scp: limit: No such file or directory", "task_level": [5]}
2015-08-13 14:03:02+0000 [-] ELIOT: {"timestamp": 1439474582.504021, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "message_type": "flocker.common.runner:run:output", "line": "scp: support: No such file or directory", "task_level": [6]}
2015-08-13 14:03:02+0000 [-] ELIOT: {"exception": "twisted.internet.error.ProcessTerminated", "task_level": [7], "action_type": "flocker.common.runner:run", "reason": "A process has ended with a probable error condition: process ended with exit code 1.", "timestamp": 1439474582.508583, "task_uuid": "b9b97ad5-daf6-4f13-b985-d0d2044f509c", "action_status": "failed"}
2015-08-13 14:03:02+0000 [-] Main loop terminated.
```

stdout and stderr are both captured by the `handle_line` function. 

Need to modify the `run` command to support separate handlers.

wallrj commented:
@jml I moved the `spawnProcess` based subprocess functions into `flocker.common` and made the necessary changes to the tests.

I think you'd better take another look before I merge this.

jml commented on flocker/acceptance/endtoend/test_diagnostics.py:
```diff
@@ -0,0 +1,103 @@
+                self.fail(
+                    'Unexpected Archive Content\n'
+                    + '\n'.join(message)
+                )
```
Why not just `assertEqual(actual_basenames, expected_basenames)`?

wallrj replied:
It was difficult to see what files were missing and unexpected. But now I've got the test working I can simplify the assertion. Done.

jml commented on flocker/node/_diagnostics.py:
```diff
@@ -0,0 +1,309 @@
+        self.name = name
+        self.version = version
+        self.service_manager = service_manager
+        self.log_exporter = log_exporter
```
:question: (optional) Make this a `PClass`

wallrj replied:
Done.

jml commented on flocker/acceptance/endtoend/test_diagnostics.py:
```diff
@@ -0,0 +1,103 @@
+                'service-status',
+                'ip-addr',
+                'hostname',
+            ])
```
Github makes it difficult to tell, but I think you asked me what my honest opinion was about this test. 

It tests a large swathe of code and makes a relatively imprecise assertion. I would generally try to avoid tests like this. However, sometimes they are the only way (or the only cost-feasible way) of getting coverage. Because Python is so easy to mess up, coverage is important. 

I guess I would prefer direct tests of the `FlockerDebugArchive.create` method. The actual script execution stuff is presumably well tested by the standard script tests, and the content of the script is so simple that the value of an integration test there is low. It's possible that direct tests of `create` would require so much set up that they too would end up being [change detector tests](http://googletesting.blogspot.co.uk/2015/01/testing-on-toilet-change-detector-tests.html). I couldn't say without trying.

Fundamentally it's a messy problem, and I think this test is OK. Does that help?

jml commented:
I think I've only requested one simple change. Please feel free to ping me directly if you'd like follow-up review or clarification.

When I started the review, one builder was failing, probably spuriously. I triggered another run of the tests there. As usual, please only merge once tests pass.

davidoliverSP2 commented on docs/using/administering/debugging.rst:
```diff
@@ -33,6 +37,78 @@ It is possible to see the available unit names, and then view the logs with ``jo
+
+     flocker-control --version
+
+* Operating system and version:
```
This step/command has a colon, while the others in this list do not. I think it looks better to have the colon before each example (rather than removing this one)

wallrj replied:
Done.

wallrj commented:
In https://github.com/ClusterHQ/flocker/pull/1774/files#r37065576 @jml wrote:
> Fundamentally it's a messy problem, and I think this test is OK. Does that help?
Yep. I guess I should have just tested that way from the start. 
Writing the acceptance test wasn't as difficult as I anticipated.
And writing a functional test or two for `FlockerDebugArchive.create` would have probably resulted in a cleaner design.

I don't think we're creating  and linking to follow up issues any more, but here are some thoughts for improvements:
 * Handle and log failure in each log export  step -- currently the script will exit if one of the subprocesses fails e.g. `docker version` fails because `docker` daemon isn't running.
 * Write incrementally from subprocess to a zip file member file -- would avoid having to compress each individual file and then tar it up in the final step. Here's a possible solution: http://stackoverflow.com/questions/297345/create-a-zip-file-from-a-generator-in-python/2734156#2734156 which we might be able to use. 
 * And at the same time use flocker.common.runner to launch subprocesses with spawnProcess in parallel.

I added the extra diagnostic output of lsblk and fdisk so putting this up for another review as requested.

jml commented on flocker/node/_diagnostics.py:
```diff
@@ -0,0 +1,320 @@
+                ['lsblk', '--all'],
+                stdout=self._open_logfile('lsblk')
+            )
+
```
Another thought for a later improvement: make each of these things data (e.g. a `Diagnostic` class with a `get` method). Really doesn't need to happen in this branch. :smile:

jml commented:
Looks like the tests are failing legitimately: http://build.clusterhq.com/builders/flocker%2Facceptance%2Faws%2Fubuntu-14.04%2Faws/builds/1715 

If the fix is simple, please address comments & merge. Otherwise, please resubmit.

wallrj commented:
Ah, damn. I assumed that all variates of `fdisk` would support the long options....they don't.

```
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.570568, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " fdisk [options] <disk>    change partition table", "task_level": [5]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.570797, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " fdisk [options] -l <disk> list partition table(s)", "task_level": [6]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.571021, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " fdisk -s <partition>      give partition size(s) in blocks", "task_level": [7]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.57124, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": "", "task_level": [8]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.571457, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": "Options:", "task_level": [9]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.571694, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -b <size>             sector size (512, 1024, 2048 or 4096)", "task_level": [10]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.571919, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -c[=<mode>]           compatible mode: 'dos' or 'nondos' (default)", "task_level": [11]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.572136, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -h                    print this help text", "task_level": [12]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.572346, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -u[=<unit>]           display units: 'cylinders' or 'sectors' (default)", "task_level": [13]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.572572, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -v                    print program version", "task_level": [14]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.572785, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -C <number>           specify the number of cylinders", "task_level": [15]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.572993, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -H <number>           specify the number of heads", "task_level": [16]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.573203, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": " -S <number>           specify the number of sectors per track", "task_level": [17]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.573411, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": "", "task_level": [18]}
2015-08-17 10:26:00+0000 [-] ELIOT: {"timestamp": 1439807160.573708, "task_uuid": "2211170a-dd39-41ef-941b-1bccebd43f09", "message_type": "flocker.common.runner:run:stderr", "line": "Unhandled Error", "task_level": [19]}
```

wallrj commented:
Ok. There are a couple of lines of `stderr` printed on the Ubuntu version which will look a little ugly for Ubuntu users:

* http://build.clusterhq.com/builders/flocker%2Facceptance%2Faws%2Fubuntu-14.04%2Faws/builds/1719/steps/trial/logs/test.log
```
2015-08-17 12:19:56+0000 [-] ELIOT: {"timestamp": 1439813996.630105, "task_uuid": "7e2dbe6d-16fb-4904-af98-0f2e7b8783be", "message_type": "flocker.common.runner:run:stderr", "line": "WARNING: No swap limit support", "task_level": [2]}
2015-08-17 12:19:56+0000 [-] ELIOT: {"timestamp": 1439813996.639077, "task_uuid": "7e2dbe6d-16fb-4904-af98-0f2e7b8783be", "message_type": "flocker.common.runner:run:stderr", "line": "Disk /dev/xvdb doesn't contain a valid partition table", "task_level": [3]}
```

Another future improvement might be to capture the stdout and stderr of each diagnostic command to separate files in the archive.

But I think I'll ignore those and merge.

wallrj commented:
Ah, I should probably add a NEWS and What's New entry for this. 
I'll do that before merging.
